rabbit-source/src/main/java/org/springframework/cloud/stream/module/rabbit/source/RabbitSource.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.amqp.rabbit.config.RetryInterceptorBuilder;
+import org.springframework.amqp.rabbit.connection.ConnectionFactory;
+import org.springframework.amqp.rabbit.listener.SimpleMessageListenerContainer;
+import org.springframework.amqp.rabbit.retry.RejectAndDontRequeueRecoverer;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.amqp.RabbitProperties;
+import org.springframework.boot.autoconfigure.amqp.RabbitProperties.Listener;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.stream.annotation.Bindings;
+import org.springframework.cloud.stream.annotation.EnableBinding;
+import org.springframework.cloud.stream.messaging.Source;
+import org.springframework.context.annotation.Bean;
+import org.springframework.integration.amqp.inbound.AmqpInboundChannelAdapter;
+import org.springframework.integration.amqp.support.DefaultAmqpHeaderMapper;
+import org.springframework.retry.interceptor.RetryOperationsInterceptor;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * A source module that receives data from RabbitMQ.
+ *
+ * @author Gary Russell
+ */
+@EnableBinding(Source.class)
+@EnableConfigurationProperties(RabbitSourceProperties.class)
+public class RabbitSource {
+
+	@Autowired
+	@Bindings(RabbitSource.class)
+	private Source channels;
+
+	@Autowired
+	private RabbitProperties rabbitProperties;
+
+	@Autowired
+	private RabbitSourceProperties properties;
+
+	@Autowired
+	private ConnectionFactory rabbitConnectionFactory;
+
+	@Bean
+	public SimpleMessageListenerContainer container() {
+		SimpleMessageListenerContainer container = new SimpleMessageListenerContainer(this.rabbitConnectionFactory);
+		Listener listenerProperties = this.rabbitProperties.getListener();
+		AcknowledgeMode acknowledgeMode = listenerProperties.getAcknowledgeMode();
+		if (acknowledgeMode != null) {
+			container.setAcknowledgeMode(acknowledgeMode);
+		}
+		Integer concurrency = listenerProperties.getConcurrency();
+		if (concurrency != null) {
+			container.setConcurrentConsumers(concurrency);
+		}
+		Integer maxConcurrency = listenerProperties.getMaxConcurrency();
+		if (maxConcurrency != null) {
+			container.setMaxConcurrentConsumers(maxConcurrency);
+		}
+		Integer prefetch = listenerProperties.getPrefetch();
+		if (prefetch != null) {
+			container.setPrefetchCount(prefetch);
+		}
+		Integer transactionSize = listenerProperties.getTransactionSize();
+		if (transactionSize != null) {
+			container.setTxSize(transactionSize);
+		}
+		container.setDefaultRequeueRejected(this.properties.getRequeue());
+		container.setChannelTransacted(this.properties.getTransacted());
+		String[] queues = this.properties.getQueues();
+		Assert.state(queues.length > 0, "At least one queue is required");
+		Assert.noNullElements(queues, "queues cannot have null elements");
```

Comment:
These tests should really be in the SMLC - they are there for `addQueueNames` but not `setQueueNames`. I'll open a JIRA over there; we can remove these later.